hazelcast-client/src/main/java/com/hazelcast/client/connection/nio/ClientConnectionManagerImpl.java

```diff
@@ -216,7 +216,15 @@ public ClientConnection getConnection(Address target) {
         if (target == null) {
             return null;
         }
-        return connections.get(target);
+        ClientConnection connection = connections.get(target);
+        if (connection != null && !connection.isAlive()) {
+            if (logger.isFinestEnabled()) {
+                logger.finest("Get a connection " + connection + " which is not alive. Destroying it.");
+            }
+            destroyConnection(connection, "Not Alive", null);
```

Comment:
I agree. I explained one scenario above, but there might be more. This whole system should be simplified.